src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -581,6 +587,118 @@ private RegexNode ReduceAtomic()
+                        bool reordered = false;
+                        for (int start = 0; start < branches.Count; start++)
+                        {
+                            // Get the node that may start our range.  If it's a one, multi, or concat of those, proceed.
+                            RegexNode startNode = branches[start];
+                            if (startNode.FindBranchOneOrMultiStart() is null)
+                            {
+                                continue;
+                            }
+
+                            // Find the contiguous range of nodes from this point that are similarly one, multie, or concat of those.
+                            int endExclusive = start + 1;
+                            while (endExclusive < branches.Count && branches[endExclusive].FindBranchOneOrMultiStart() is not null)
+                            {
+                                endExclusive++;
+                            }
+
+                            // If there's at least 3, there may be something to reorder (we won't reorder anything
+                            // before the starting position, and so only 2 items is considered ordered).
+                            if (endExclusive - start >= 3)
+                            {
+                                int compare = start;
+                                while (compare < endExclusive)
+                                {
+                                    // Get the starting character
+                                    char c = branches[compare].FindBranchOneOrMultiStart()!.FirstCharOfOneOrMulti();
+
+                                    // Move compare to point to the last branch that has the same starting value.
+                                    while (compare < endExclusive && branches[compare].FindBranchOneOrMultiStart()!.FirstCharOfOneOrMulti() == c)
+                                    {
+                                        compare++;
+                                    }
+
+                                    // Compare now points to the first node that doesn't match the starting node.
+                                    // If we've walked off our range, there's nothing left to reorder.
+                                    if (compare < endExclusive)
+                                    {
+                                        // There may be something to reorder.  See if there are any other nodes that begin with the same character.
+                                        for (int next = compare + 1; next < endExclusive; next++)
+                                        {
+                                            RegexNode nextChild = branches[next];
+                                            if (nextChild.FindBranchOneOrMultiStart()!.FirstCharOfOneOrMulti() == c)
+                                            {
+                                                for (int move = next; move > compare; move--)
```

Comment:
Easier, yes, but RemoveAt will shift everything down and then Insert will shift everything back up.  Figured it'd be better to do it manually, just in case the list were really long.  But... given that it's unlikely these are really big, I can switch back to just using RemoveAt/Insert.